### PR TITLE
[SPARK-14836][Yarn] Zip all the jars before uploading to distributed cache

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

<copy form JIRA>

Currently if neither `spark.yarn.jars` nor `spark.yarn.archive` is set (by default), Spark on yarn code will upload all the jars in the folder separately into distributed cache, this is quite time consuming, and very verbose, instead of upload jars separately into distributed cache, here changes to zip all the jars first, and then put into distributed cache.

This will significantly improve the speed of starting time.
## How was this patch tested?

Unit test and local integrated test is done.

Verified with SparkPi both in spark cluster and client mode.

### CODE REVIEW

**Comment by jerryshao** [comment]
CC @vanzin , please help to review, thanks a lot.

**Comment by lianhuiwang** [comment]
LGTM

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -492,15 +492,30 @@ private[spark] class Client(
 
         case None =>
           // No configuration, so fall back to uploading local jar files.
-          logWarning(s"Neither ${SPARK_JARS.key} nor ${SPARK_ARCHIVE.key} is set, falling back " +
+          logInfo(s"Neither ${SPARK_JARS.key} nor ${SPARK_ARCHIVE.key} is set, falling back " +
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I really think this should be a warning. Falling back into this mode means your Spark app's startup time is much slower. People really should be aware that they're falling into this case, and a warning helps more than an info log, even if a tiny bit.

**Comment by vanzin** [comment]
LGTM aside from the log change.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (jerryshao) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: OK, I will change it back.

**Comment by vanzin** [comment]
Merging to master.

**Pull request closed without merging**